cluster-autoscaler/simulator/scheduling/hinting_simulator.go

```diff
@@ -44,53 +50,73 @@ func NewHintingSimulator(predicateChecker predicatechecker.PredicateChecker) *Hi
 // Each node is considered acceptable iff isNodeAcceptable() returns true.
-// Returns a list of nodes that were chosen or an error if scheduling any of the pods failed.
+// Returns a list of scheduled nodes or an error if an unexpected error occurs.
+// If the breakOnFailure is set to true, the function will stop scheduling attempts
+// after the first failure. This is useful if all provided pods need to be scheduled.
 // Note: this function does not fork clusterSnapshot: this has to be done by the caller.
-func (s *HintingSimulator) TrySchedulePods(clusterSnapshot clustersnapshot.ClusterSnapshot, pods []*apiv1.Pod, isNodeAcceptable func(string) bool) ([]string, error) {
-	var nodeNames []string
+func (s *HintingSimulator) TrySchedulePods(clusterSnapshot clustersnapshot.ClusterSnapshot, pods []*apiv1.Pod, isNodeAcceptable func(*schedulerframework.NodeInfo) bool, breakOnFailure bool) ([]Status, error) {
+	s.similarPods = NewSimilarPodsScheduling()
+
+	var statuses []Status
 	for _, pod := range pods {
 		klog.V(5).Infof("Looking for place for %s/%s", pod.Namespace, pod.Name)
 		nodeName, err := s.findNodeWithHints(clusterSnapshot, pod, isNodeAcceptable)
 		if err != nil {
 			return nil, err
 		}
+
 		if nodeName == "" {
-			nodeName, err = s.schedulePod(clusterSnapshot, pod, isNodeAcceptable)
+			nodeName, err = s.findNode(clusterSnapshot, pod, isNodeAcceptable)
 			if err != nil {
 				return nil, err
 			}
 		}
-		nodeNames = append(nodeNames, nodeName)
+
+		if nodeName != "" {
+			klog.V(4).Infof("Pod %s/%s can be moved to %s", pod.Namespace, pod.Name, nodeName)
+			if err := clusterSnapshot.AddPod(pod, nodeName); err != nil {
+				return nil, fmt.Errorf("simulating scheduling of %s/%s to %s return error; %v", pod.Namespace, pod.Name, nodeName, err)
+			}
+			statuses = append(statuses, Status{Pod: pod, NodeName: nodeName})
+		} else if breakOnFailure {
+			break
+		}
 	}
-	return nodeNames, nil
+	return statuses, nil
 }
 
-func (s *HintingSimulator) findNodeWithHints(clusterSnapshot clustersnapshot.ClusterSnapshot, pod *apiv1.Pod, isNodeAcceptable func(string) bool) (string, error) {
+func (s *HintingSimulator) findNodeWithHints(clusterSnapshot clustersnapshot.ClusterSnapshot, pod *apiv1.Pod, isNodeAcceptable func(*schedulerframework.NodeInfo) bool) (string, error) {
 	hk := HintKeyFromPod(pod)
-	if hintedNode, hasHint := s.hints.Get(hk); hasHint && isNodeAcceptable(hintedNode) {
+	if hintedNode, hasHint := s.hints.Get(hk); hasHint {
 		if err := s.predicateChecker.CheckPredicates(clusterSnapshot, pod, hintedNode); err == nil {
-			klog.V(4).Infof("Pod %s/%s can be moved to %s", pod.Namespace, pod.Name, hintedNode)
-			if err := clusterSnapshot.AddPod(pod, hintedNode); err != nil {
-				return "", fmt.Errorf("Simulating scheduling of %s/%s to %s return error; %v", pod.Namespace, pod.Name, hintedNode, err)
-			}
 			s.hints.Set(hk, hintedNode)
-			return hintedNode, nil
+
+			nodeInfo, err := clusterSnapshot.NodeInfos().Get(hintedNode)
```

Comment:
Assume the scenario when pod A is able to get scheduled only on node B. We detect it once and cache it. Then somebody runs the algorithm again, this time with node B being non-acceptable. This can happen e.g. in scale-down, when non-acceptable nodes are different each run.
If we just ignored the hint because node B is non-acceptable, then we would remove this data from cache. This is not ideal - consecutive call can have B as acceptable, but the hint will already be deleted.

To recap - thanks to this single predicate run we can potentially avoid O(N) predicate checks in the future. This is a small optimalization, but can potentially help.

Reply:
Ack, makes sense